File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] res = new int[nums.length - k + 1];
        Deque<Integer> dq = new ArrayDeque<>();
        int l = 0, r = 0;
        while (r < nums.length) {
            while (!dq.isEmpty() && nums[dq.peekLast()] < nums[r]) {
                dq.pollLast();
            }
            dq.offer(r);
            if (l > dq.peekFirst()) dq.pollFirst();
            if (r + 1 >= k) {
                res[l] = nums[dq.peekFirst()];
                l++;
            }
            r++;
        }
        return res;
    }
}

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.kt
class Solution {
    fun maxSlidingWindow(nums: IntArray, k: Int): IntArray {
        val res = IntArray(nums.size - k + 1)
        val dq = ArrayDeque<Int>()
        var l = 0
        var r = 0
        while (r < nums.size) {
            while (!dq.isEmpty() && nums[dq.last()] < nums[r]) dq.removeLast()
            dq.addLast(r)
            if (l > dq.first()) dq.removeFirst()
            if (r + 1 >= k) {
                res[l] = nums[dq.first()]
                l++
            }
            r++
        }
        return res
    }
}